File: opcua-webservice-node/src/server/addressspace/nodes/UaDefinitionNode.ts
import { UaNodeId, UaLocalizedText, UaVariant, UaError, makeUaStatusCode } from "opcua-webapi-ts"; // Adjust paths as necessary
import { UaNode } from './UaNode';
import { UaReference } from './UaReference';
import { Attributes, StatusCodes } from "opcua-webapi";
import { UaReferenceTypes } from "./builtin";

export abstract class UaDefinitionNode extends UaNode {
    private readonly _isAbstract: boolean;
    private _parentType: UaDefinitionNode | null;

    constructor(
        nodeId: UaNodeId,
        browseName: string,
        displayName: UaLocalizedText,
        isAbstract: boolean
    ) {
        super(nodeId, browseName, displayName);
        this._isAbstract = isAbstract;
        this._parentType = null;
    }

    public get isAbstract(): boolean {
        return this._isAbstract;
    }
    
    public get parentType(): UaDefinitionNode | null {
        return this._parentType;
    }

    public isSubtypeOf(typeId: UaNodeId): boolean {
        if (this.nodeId.equal(typeId)) return true;
        if (this._parentType === null) return false;
        return this._parentType.isSubtypeOf(typeId);
    }

    protected setParentType(parentType: UaDefinitionNode): void {
        if (this.nodeClass !== parentType.nodeClass) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadNodeClassInvalid));
        }
        if (this._parentType !== null) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadAlreadyExists));
        }

        this._parentType = parentType;
        this.addReference(new UaReference(parentType, UaReferenceTypes.HasSubtype, false));
        parentType.addReference(new UaReference(this, UaReferenceTypes.HasSubtype, true));
    }

    public getAttribute(attributeId: number): UaVariant {
        if (Attributes.IsAbstract === attributeId) {
            return UaVariant.boolean(this._isAbstract);
        }

        return super.getAttribute(attributeId);
    }
}

File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaDefinitionNode.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaReferenceTypes;
import org.springframework.lang.Nullable;

public abstract class UaDefinitionNode extends UaNode {
    private final boolean isAbstract;
    private @Nullable UaDefinitionNode parentType;

    public UaDefinitionNode(
            NodeId nodeId,
            String browseName,
            LocalizedText displayName,
            boolean isAbstract)
    {
        super(nodeId, browseName, displayName);
        this.isAbstract = isAbstract;
        parentType = null;
    }

    public boolean isAbstract() {
        return isAbstract;
    }

    @Nullable
    public UaDefinitionNode parentType() {
        return parentType;
    }

    public boolean isSubtypeOf(NodeId typeId)
    {
        if (typeId.equals(nodeId())) return true;
        if (null == parentType) return false;
        return parentType.isSubtypeOf(typeId);
    }

    public @Nullable UaDefinitionNode getParentType()
    {
        return parentType;
    }

    public void setParentType(UaDefinitionNode parentType) throws UaRuntimeException {
        if (nodeClass() != parentType.nodeClass()) throw new UaRuntimeException(StatusCodes.Bad_NodeClassInvalid);
        if (null != this.parentType) throw new UaRuntimeException(StatusCodes.Bad_AlreadyExists);

        this.parentType = parentType;
        this.addReference(new UaReference(parentType, UaReferenceTypes.HasSubType,false));
        parentType.addReference(new UaReference(this, UaReferenceTypes.HasSubType,true));
    }

    @Override
    public Variant getAttribute(int attributeId)
    {
        if (AttributeId.IsAbstract.id() == attributeId)
        {
            return new Variant(isAbstract);
        }

        return super.getAttribute(attributeId);
    }
}
